experimental/devices/epd/epd.go

```diff
@@ -0,0 +1,463 @@
+}
+
+// Dev is an open handle to the display controller.
+type Dev struct {
+	// Communication
+	c    conn.Conn
+	dc   gpio.PinOut
+	cs   gpio.PinOut
+	rst  gpio.PinOut
+	busy gpio.PinIO
+
+	// Display size controlled by the e-paper display.
+	rect image.Rectangle
+
+	lt     LUTType
+	config *DisplayConfig
+}
+
+func (d *Dev) String() string {
+	return fmt.Sprintf("epd.Dev{%s, %s, %s}", d.c, d.dc, d.rect.Max)
+}
+
+// ColorModel implements display.Drawer.
+// It is a one bit color model, as implemented by image1bit.Bit.
+func (d *Dev) ColorModel() color.Model {
+	return image1bit.BitModel
+}
+
+// Bounds implements display.Drawer. Min is guaranteed to be {0, 0}.
+func (d *Dev) Bounds() image.Rectangle {
+	return d.rect
+}
+
+// Draw implements display.Drawer.
+func (d *Dev) Draw(r image.Rectangle, src image.Image, sp image.Point) error {
+	xStart := sp.X
+	yStart := sp.Y
+	imageW := r.Dx() & 0xF8
+	imageH := r.Dy()
+	w := d.rect.Dx()
+	h := d.rect.Dy()
+
+	xEnd := xStart + imageW - 1
+	if xStart+imageW >= w {
+		xEnd = w - 1
+	}
+
+	yEnd := yStart + imageH - 1
+	if yStart+imageH >= h {
+		yEnd = h - 1
+	}
+
+	d.setMemoryArea(xStart, yStart, xEnd, yEnd)
+
+	next := image1bit.NewVerticalLSB(d.rect)
+	draw.Src.Draw(next, r, src, sp)
+	var byteToSend byte = 0x00
+	for y := yStart; y < yEnd+1; y++ {
+		d.setMemoryPointer(xStart, y)
+		if err := d.sendCommand([]byte{byte(WriteRAM)}); err != nil {
+			return err
+		}
+		for x := xStart; x < xEnd+1; x++ {
+			bit := next.BitAt(x-xStart, y-yStart)
+			if bit {
+				byteToSend |= 0x80 >> (uint32(x) % 8)
+			}
+			if x%8 == 7 {
+				if err := d.sendData([]byte{byte(byteToSend)}); err != nil {
+					return err
+				}
+				byteToSend = 0x00
+			}
+		}
+	}
+
+	return nil
+}
+
+// ClearFrameMemory clear the frame memory with the specified color.
+// this won't update the display.
+func (d *Dev) ClearFrameMemory(color byte) error {
+	w := d.rect.Dx()
+	h := d.rect.Dy()
+	d.setMemoryArea(0, 0, w-1, h-1)
+	d.setMemoryPointer(0, 0)
+	if err := d.sendCommand([]byte{byte(WriteRAM)}); err != nil {
+		return err
+	}
+	// send the color data
+	for i := 0; i < (w / 8 * h); i++ {
+		if err := d.sendData([]byte{color}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DisplayFrame update the display
+// there are 2 memory areas embedded in the e-paper display
+// but once this function is called,
+// the next action of SetFrameMemory or ClearFrame will
+// set the other memory area.
+func (d *Dev) DisplayFrame() error {
+	if err := d.sendCommand([]byte{byte(DisplayUpdateControl2)}); err != nil {
+		return err
+	}
+
+	if err := d.sendData([]byte{byte(0xC4)}); err != nil {
+		return err
+	}
+
+	if err := d.sendCommand([]byte{byte(MasterActivation)}); err != nil {
+		return err
+	}
+
+	if err := d.sendCommand([]byte{byte(TerminateFrameReadWrite)}); err != nil {
+		return err
+	}
+
+	d.waitUntilIdle()
+	return nil
+}
+
+// Halt turns off the display.
+func (d *Dev) Halt() error {
+	return d.Sleep()
```

Comment:
I don't think sleep clears the display?

Reply:
I'm not sure what is the best thing to do here. Indeed sending the device to sleep was not the best option, but now I'm clearing the screen. In what moment is recommended to make the call to `Halt`, because in the case of the ePaper display, the image will stay there even after powering off the display, so I don't think that the user wants to clear the display.